src/VisualStudio/Core/Test.Next/Services/LspDiagnosticsTests.cs

```diff
@@ -0,0 +1,470 @@
+using Microsoft.CodeAnalysis.Test.Utilities;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.LanguageServices.Implementation.LanguageService;
+using Moq;
+using Nerdbank.Streams;
+using Newtonsoft.Json.Linq;
+using Roslyn.Test.Utilities;
+using Roslyn.Utilities;
+using StreamJsonRpc;
+using Xunit;
+using LSP = Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Roslyn.VisualStudio.Next.UnitTests.Services
+{
+    [UseExportProvider]
+    public class LspDiagnosticsTests : AbstractLanguageServerProtocolTests
+    {
+        [Fact]
+        public async Task AddDiagnosticTestAsync()
+        {
+            using var workspace = CreateTestWorkspace("", out _);
+            var document = workspace.CurrentSolution.Projects.First().Documents.First();
+
+            var diagnosticsMock = new Mock<IDiagnosticService>();
+            // Create a mock that returns a diagnostic for the document.
+            SetupMockWithDiagnostics(diagnosticsMock, document.Id, await CreateMockDiagnosticDataAsync(document, "id").ConfigureAwait(false));
+
+            // Publish one document change diagnostic notification ->
+            // 1.  doc1 with id.
+            //
+            // We expect one publish diagnostic notification ->
+            // 1.  from doc1 with id.
+            var results = await RunPublishDiagnosticsAsync(workspace, diagnosticsMock.Object, 1, document).ConfigureAwait(false);
+
+            // We expect one publish notification for the diagnostic on the document.
+            Assert.Equal(new Uri(document.FilePath), results.Single().Uri);
+            Assert.Equal("id", results.Single().Diagnostics.Single().Code);
+        }
+
+        [Fact]
+        public async Task AddDiagnosticWithMappedFilesTestAsync()
+        {
+            using var workspace = CreateTestWorkspace("", out _);
+            var document = workspace.CurrentSolution.Projects.First().Documents.First();
+
+            var diagnosticsMock = new Mock<IDiagnosticService>();
+            // Create two mapped diagnostics for the document.
+            SetupMockWithDiagnostics(diagnosticsMock, document.Id,
+                await CreateMockDiagnosticDatasWithMappedLocationAsync(document, ("id1", document.FilePath + "m1"), ("id2", document.FilePath + "m2")).ConfigureAwait(false));
+
+            // Publish one document change diagnostic notification ->
+            // 1.  doc1 with id1 = mapped file m1 and id2 = mapped file m2.
+            //
+            // We expect two publish diagnostic notifications ->
+            // 1.  from m1 with id1 (from 1 above).
+            // 2.  from m2 with id2 (from 1 above).
+            var results = await RunPublishDiagnosticsAsync(workspace, diagnosticsMock.Object, expectedNumberOfCallbacks: 2, document).ConfigureAwait(false);
+
+            Assert.Equal(2, results.Count);
+            Assert.Equal("id1", results.Single(p => p.Uri == new Uri(document.FilePath + "m1")).Diagnostics.Single().Code);
```

Comment:
Work having this return a dictionary or something so you don't have to write the whole `results.Single(p => p.Uri == new Uri(` thing multiple times?

Reply:
I actually started with dictionary, but it was a bit of a struggle.  With a dictionary, I can only verify the end result (after all publish notifications are complete) since we'd be replacing diagnostics for a particular URI as we publish.  I found it much easier to write tests (and I think more correct) when I can verify that each incremental publish notification produces the correct output.